## Odds to die on Titanic and logistic regression with R

`So you'd like to calculte your odds to survive on Titanic disaster?
Well, before that, let's start with it we'll learn a bit about Logistic regression with R language.
We are basically going to analyze data about people who died and who survived (their gender, age, in  which class they were travelling...). With that, we'll have a simple equation to calculate your odds to survive, based on that characteristics.
Let's do that?`

`First thing we do is importing our csv data about all those people and check our data.`

In [14]:
DataTitanic <- read.csv2("Data Titanic.csv")
DataTitanic

Passageiro Sobrevivente Classe
1    1         0            3     
2    2         1            1     
3    3         1            3     
4    4         1            1     
5    5         0            3     
6    7         0            1     
7    8         0            3     
8    9         1            3     
9   10         1            2     
10  11         1            3     
11  12         1            1     
12  13         0            3     
13  14         0            3     
14  15         0            3     
15  16         1            2     
16  17         0            3     
17  19         0            3     
18  21         0            2     
19  22         1            2     
20  23         1            3     
21  24         1            1     
22  25         0            3     
23  26         1            3     
24  28         0            1     
25  31         0            1     
26  34         0            2     
27  35         0            1     
28  36         0            1     
29  38         0            3     
30  39         0            3     
... ...        ...          ...   
861 NA         NA           NA    
862 NA         NA           NA    
863 NA         NA           NA    
864 NA         NA           NA    
865 NA         NA           NA    
866 NA         NA           NA    
867 NA         NA           NA    
868 NA         NA           NA    
869 NA         NA           NA    
870 NA         NA           NA    
871 NA         NA           NA    
872 NA         NA           NA    
873 NA         NA           NA    
874 NA         NA           NA    
875 NA         NA           NA    
876 NA         NA           NA    
877 NA         NA           NA    
878 NA         NA           NA    
879 NA         NA           NA    
880 NA         NA           NA    
881 NA         NA           NA    
882 NA         NA           NA    
883 NA         NA           NA    
884 NA         NA           NA    
885 NA         NA           NA    
886 NA         NA           NA    
887 NA         NA           NA    
888 NA         NA           NA    
889 NA         NA           NA    
890 NA         NA           NA    
    Nome                                                      Sexo      Idade
1   Braund, Mr. Owen Harris                                   masculino 22   
2   Cumings, Mrs. John Bradley (Florence Briggs Thayer)       feminino  38   
3   Heikkinen, Miss. Laina                                    feminino  26   
4   Futrelle, Mrs. Jacques Heath (Lily May Peel)              feminino  35   
5   Allen, Mr. William Henry                                  masculino 35   
6   McCarthy, Mr. Timothy J                                   masculino 54   
7   Palsson, Master. Gosta Leonard                            masculino  2   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         feminino  27   
9   Nasser, Mrs. Nicholas (Adele Achem)                       feminino  14   
10  Sandstrom, Miss. Marguerite Rut                           feminino   4   
11  Bonnell, Miss. Elizabeth                                  feminino  58   
12  Saundercock, Mr. William Henry                            masculino 20   
13  Andersson, Mr. Anders Johan                               masculino 39   
14  Vestrom, Miss. Hulda Amanda Adolfina                      feminino  14   
15  Hewlett, Mrs. (Mary D Kingcome)                           feminino  55   
16  Rice, Master. Eugene                                      masculino  2   
17  Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)   feminino  31   
18  Fynney, Mr. Joseph J                                      masculino 35   
19  Beesley, Mr. Lawrence                                     masculino 34   
20  McGowan, Miss. Anna "Annie"                               feminino  15   
21  Sloper, Mr. William Thompson                              masculino 28   
22  Palsson, Miss. Torborg Danira                             feminino   8   
23  Asplund, Mrs. Carl Oscar (Selma Augu

`After, we'll get our logitic model with glm function.
Let's check first whether your gender will influence on your survive odd (if you watched the movie, you already know the answer). As we are talking about a chance of success (survive) and failure (die), we are going to work with a binomial error distribution.`

In [20]:
model <- glm(Sobrevivente~Sexo, family = binomial(link="logit"))
summary(model)


Call:
glm(formula = Sobrevivente ~ Sexo, family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6767  -0.6779  -0.6779   0.7501   1.7795  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept)     1.1243     0.1439   7.814 5.52e-15 ***
Sexomasculino  -2.4778     0.1850 -13.392  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 750.70  on 712  degrees of freedom
  (176 observations deleted due to missingness)
AIC: 754.7

Number of Fisher Scoring iterations: 4


`We have some information now to analyze.`
* On intercept, we see the Estimate Std ($\beta_0 = 1.1243$).
* Also, We have a male gender Estimate Std. ($\beta_1 = -2.4778$). The value of $e^{\beta_1} = 0.084 = 8.4\%$ indicates how much the odds radio changes when we go from a category to other. (Whaaaat?)
* In other words, when someone from the female group becomes male, they have $1 - 0.084 = 0.916 = 92\%$ more chance of dying.
* Notice that, with a negative $\beta_1$, the chances of failure (not surviving) are greater with men than women.
* We can also calculate: $1/e^{\beta_1} = 11.91$. It means that the female group is 11 times more successful than the male group. In other words, you'd have 11 more chances to survive if you were a woman on Titanic.

`We can also analyze whether our model is significant.`
* We see that P-value ($Pr(>|z|)$) are very low for $\beta_0$ and $\beta_1$.
* Usually, when P-value < 0.05, we say that the variable is significant.
* Also, the stars next to the P-value mean their value range. Three stars ($***$) means a value between 0 and 0.1%.

`We can also analyze it based on age information.
The question here is: "if I'm older or younger, my chances to survive will increase?"
To answer this, we can create a model and check the variables.`

In [22]:
model2 <- glm(Sobrevivente~Idade, family = binomial(link="logit"))
summary(model2)


Call:
glm(formula = Sobrevivente ~ Idade, family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.1488  -1.0361  -0.9544   1.3159   1.5908  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)  
(Intercept) -0.05672    0.17358  -0.327   0.7438  
Idade       -0.01096    0.00533  -2.057   0.0397 *
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 960.23  on 712  degrees of freedom
  (176 observations deleted due to missingness)
AIC: 964.23

Number of Fisher Scoring iterations: 4


`First thing to notice is that $\beta_1$ in this case is negative, meaning that when age increases, your chances of success decrease.`

Calculating $e^{-0.01096} = 0.99$, we'll know that, for one year that you are older, your chance of failure will increase 0.99%.


`Another analysis can be: "What if I decide to go to a higher class? Will my odds to survive increase?"
Once again, if you watched the movie, you already know the answer...`

In [27]:
model3 <- glm(Sobrevivente~Classe, family = binomial(link="logit"));
summary(model3)


Call:
glm(formula = Sobrevivente ~ Classe, family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.4892  -0.7530  -0.7530   0.8948   1.6727  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  1.62053    0.22852   7.091 1.33e-12 ***
Classe      -0.91199    0.09854  -9.255  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 870.73  on 712  degrees of freedom
  (176 observations deleted due to missingness)
AIC: 874.73

Number of Fisher Scoring iterations: 4


We see here that $\beta_1 = -0.91199$. It means, when you increase your class, your odds to survive are *lower* in $e^{-0.912} = 0.4 = 40\%$.
What??

Well, remember that increase your class means that the number representing it will decrease (1st class is higher than the 2nd, which is higher than the 3rd), but R doesn't know that yet. He thinks that class is only an integer.

In [29]:
class(Classe)

[1] "integer"

`One way to make it understand how people are divided on ocean liners based upon on how much they could pay for it is to convert the class information to a factor. To make our life easier, let's create a new variable called ClassCat and include it on our data base.`

In [40]:
DataTitanic$ClasseCat <- as.factor(Classe)
class(DataTitanic$ClasseCat)

[1] "factor"

In [44]:
model4 <- glm(Sobrevivente~DataTitanic$ClasseCat, family = binomial(link="logit"))
summary(model4)


Call:
glm(formula = Sobrevivente ~ DataTitanic$ClasseCat, family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.4607  -0.7399  -0.7399   0.9184   1.6908  

Coefficients:
                       Estimate Std. Error z value Pr(>|z|)    
(Intercept)              0.6451     0.1543   4.180 2.92e-05 ***
DataTitanic$ClasseCat2  -0.7261     0.2168  -3.350 0.000808 ***
DataTitanic$ClasseCat3  -1.8009     0.1982  -9.086  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 869.81  on 711  degrees of freedom
  (176 observations deleted due to missingness)
AIC: 875.81

Number of Fisher Scoring iterations: 4


Now, we have a beta value for each class
* classe1: exp (0.6451) = 1.906
   * 1.9x more chances to survive compared to other classes
* class2: exp (-0.7261) = 0.48
   * 52% more likely to die on 2nd class compared to 1st class
* class3: exp (-1.8009) = 0.165
   * 83% more likely to die on 3rd class compared to 1st class

`With this, let's create our full model with all important characterists.`

In [45]:
model5 <- glm(Sobrevivente~Sexo+Idade+Irmãos+Pais+DataTitanic$ClasseCat, family = binomial(link="logit"))
summary(model5)


Call:
glm(formula = Sobrevivente ~ Sexo + Idade + Irmãos + Pais + DataTitanic$ClasseCat, 
    family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.7723  -0.6353  -0.3869   0.6327   2.4513  

Coefficients:
                        Estimate Std. Error z value Pr(>|z|)    
(Intercept)             4.356474   0.455990   9.554  < 2e-16 ***
Sexomasculino          -2.642267   0.219749 -12.024  < 2e-16 ***
Idade                  -0.044836   0.008229  -5.448 5.09e-08 ***
Irmãos                 -0.368280   0.126827  -2.904  0.00369 ** 
Pais                   -0.038607   0.119601  -0.323  0.74685    
DataTitanic$ClasseCat2 -1.415376   0.284712  -4.971 6.65e-07 ***
DataTitanic$ClasseCat3 -2.650169   0.285724  -9.275  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 636.46  on 707  degree

Quickly checking the asteristics we can know that the variable related to parents is not really important for all model.
Let's remove this information from our final model.

In [48]:
model6 <- glm(Sobrevivente~Sexo+Idade+Irmãos+DataTitanic$ClasseCat, family = binomial(link="logit"))
summary(model6)


Call:
glm(formula = Sobrevivente ~ Sexo + Idade + Irmãos + DataTitanic$ClasseCat, 
    family = binomial(link = "logit"))

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.7876  -0.6417  -0.3864   0.6261   2.4539  

Coefficients:
                        Estimate Std. Error z value Pr(>|z|)    
(Intercept)             4.334201   0.450700   9.617  < 2e-16 ***
Sexomasculino          -2.627679   0.214771 -12.235  < 2e-16 ***
Idade                  -0.044760   0.008225  -5.442 5.27e-08 ***
Irmãos                 -0.380190   0.121516  -3.129  0.00176 ** 
DataTitanic$ClasseCat2 -1.414360   0.284727  -4.967 6.78e-07 ***
DataTitanic$ClasseCat3 -2.652618   0.285832  -9.280  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 964.52  on 713  degrees of freedom
Residual deviance: 636.56  on 708  degrees of freedom
  (176 observations deleted due to missingness)
AIC: 648.56

`Ok, so with this we can create our fuction to calculate your odds to survive on Titanic.`

In [65]:
odds2survive <- function(gender, age, brothersister, class) {
    beta <- 4.334201 -0.044760 * age + -0.380190 * brothersister
    if (gender == 'M'){
        beta <- beta - 2.627679
    }
    if (class == 2){
        beta <- beta - 1.414360
    }
    if (class == 3){
        beta <- beta - 2.652618
    }
    odds <- exp(beta)/(1 + exp(beta))
    
    cat(sprintf("Your odds to survive on Titanic are: %f", odds))
}

`Let's check what's the odds for a 17 years-old Rose to survive:`

In [71]:
odds2survive('F', 17, 0, 1)

Your odds to survive on Titanic are: 0.972702

`Now, let's check Jack's odds:`

In [72]:
odds2survive('M', 20, 0, 3)

Your odds to survive on Titanic are: 0.136898

`Well, may the odds be ever in your favor!`